overlord/hookstate/ctlcmd/refresh.go

```diff
@@ -122,33 +125,71 @@ type updateDetails struct {
 	restart bool
 }
 
-func getUpdateDetails(context *hookstate.Context) *updateDetails {
+// refreshCandidate is a subset of refreshCandidate defined by snapstate and
+// stored in "refresh-candidates".
+type refreshCandidate struct {
+	Channel     string         `json:"channel,omitempty"`
+	Version     string         `json:"version,omitempty"`
+	SideInfo    *snap.SideInfo `json:"side-info,omitempty"`
+	InstanceKey string         `json:"instance-key,omitempty"`
+}
+
+func getUpdateDetails(context *hookstate.Context) (*updateDetails, error) {
 	context.Lock()
 	defer context.Unlock()
 
 	if context.IsEphemeral() {
 		// TODO: support ephemeral context
-		return nil
+		return nil, nil
 	}
 
 	var base, restart bool
 	context.Get("base", &base)
 	context.Get("restart", &restart)
 
-	// TODO: get revision, version etc. from refresh-candidates.
+	var candidates []*refreshCandidate
+	st := context.State()
+	if err := st.Get("refresh-candidates", &candidates); err != nil {
+		return nil, err
+	}
 
 	up := updateDetails{
 		base:    base,
 		restart: restart,
 	}
-	return &up
+
+	// try to find revision/version/channel info from refresh-candidates; it
+	// may be missing if the hook is called for snap that is just affected by
+	// refresh but not refreshed itself, in such case this data is not
+	// displayed.
+	for _, cand := range candidates {
+		instanceName := snap.InstanceName(cand.SideInfo.RealName, cand.InstanceKey)
```

Comment:
if candidates was a map with names we wouldn't need to do this

Reply:
Agreed, changed to map per your suggestion from the other PR.